docs: Breaking change — Some Microsoft.Extensions packages in .NET 11 shared framework (#53565)
dotnet/docs: docs/core/compatibility/extensions/11/extensions-in-shared-framework.md

---
title: "Breaking change: Some Microsoft.Extensions packages included in shared framework"
description: "Learn about the breaking change in .NET 11 where nine Microsoft.Extensions.* libraries are now part of the base shared framework."
ms.date: 05/05/2026
ai-usage: ai-assisted
---

# Some Microsoft.Extensions packages included in shared framework

> To reduce application size, simplify package dependencies, and streamline servicing, .NET 11 includes nine `Microsoft.Extensions.*` libraries in the base shared framework. Projects that explicitly reference these packages receive build warning [NU1510](/nuget/reference/errors-and-warnings/nu1510). You can resolve the warning by removing the `PackageReference`. If you depend on an older version of these packages, upgrading to the .NET 11 version might expose previously undocumented breaking changes introduced between older versions and .NET 11.

Reviewer: > upgrading to the .NET 11 version might expose previously undocumented breaking changes introduced between older versions and .NET 11

I believe these breaking changes were documented (the ones listed at the end of this article).

> To reduce application size, simplify package dependencies, and streamline servicing, .NET 11 includes nine `Microsoft.Extensions.*` libraries in the base shared framework. Projects that explicitly reference these packages receive build warning [NU1510](/nuget/reference/errors-and-warnings/nu1510). You can resolve the warning by removing the `PackageReference`. If you depend on an older version of these packages, upgrading to the .NET 11 version might expose previously undocumented breaking changes introduced between older versions and .NET 11.

Reviewer 2: ```suggestion
To reduce application size, simplify package dependencies, and streamline servicing, .NET 11 includes nine `Microsoft.Extensions.*` libraries in the base shared framework. Projects that explicitly reference these packages receive build warning [NU1510](/nuget/reference/errors-and-warnings/nu1510). You can resolve the warning by removing the `PackageReference`. If you depend on an older version of these packages, upgrading to the .NET 11 version might expose breaking changes introduced between older versions and .NET 11.
```

## Version introduced

.NET 11 Preview 4

## Previous behavior

> Previously, the following `Microsoft.Extensions.*` libraries weren't part of the .NET base shared framework. Projects that needed them required explicit `PackageReference` entries, and the build process copied the assemblies to the output folder:

Reviewer: > the .NET base shared framework

Is it worth mentioning that you get the shared framework by using `Sdk="Microsoft.NET.Sdk"`, or similarly referencing any of [the other SDKs](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/overview#available-sdks)?

Reviewer 2: ```suggestion
Previously, the following `Microsoft.Extensions.*` libraries weren't part of the .NET shared framework. (You get the shared framework by using `Sdk="Microsoft.NET.Sdk"`, or similarly referencing any of [the other .NET SDKs](../../../project-sdk/overview.md#available-sdks).) Projects that needed them required explicit `PackageReference` entries, and the build process copied the assemblies to the output folder:
```

---
title: "Breaking change: Microsoft.Extensions.* packages included in shared framework"
description: "Learn about the breaking change in .NET 11 where nine Microsoft.Extensions.* libraries are now part of the base shared framework."
ms.date: 05/05/2026
ai-usage: ai-assisted
---

> # Microsoft.Extensions.* packages included in shared framework

Reviewer 2: ```suggestion
# Some Microsoft.Extensions packages included in shared framework
```

## Version introduced

.NET 11 Preview 4

## Previous behavior

Previously, the following `Microsoft.Extensions.*` libraries weren't part of the .NET base shared framework. Projects that needed them required explicit `PackageReference` entries, and the build process copied the assemblies to the output folder:

- `Microsoft.Extensions.Caching.Abstractions`
- `Microsoft.Extensions.Configuration.Abstractions`
- `Microsoft.Extensions.DependencyInjection.Abstractions`
- `Microsoft.Extensions.Diagnostics.Abstractions`
- `Microsoft.Extensions.FileProviders.Abstractions`
- `Microsoft.Extensions.Hosting.Abstractions`
- `Microsoft.Extensions.Logging.Abstractions`
- `Microsoft.Extensions.Options`

> - `Microsoft.Extensions.Primitives`

Reviewer 2: Please make these links to the NuGet packages.

## New behavior

Starting in .NET 11, these nine libraries are part of the .NET base shared framework:

- You don't need a `PackageReference` for these libraries when you target `net11.0` or later.
- If you reference these packages explicitly, you receive build warning [NU1510](/nuget/reference/errors-and-warnings/nu1510).
- To resolve NU1510, remove the `PackageReference`. The library is always available through the framework.
- These assemblies are no longer copied to the output folder.
- In rare cases, the additional APIs in the default load set might cause name or type conflicts. To resolve a conflict, add more explicit `using` directives, use an alias, or use a fully qualified type name.

> - When you target multiple frameworks (for example, `<TargetFrameworks>net10.0;net11.0</TargetFrameworks>`), the upgrade to the .NET 11 version of these libraries on the `net11.0` TFM is silent—NU1510 isn't produced in that case.

Reviewer 2: ```suggestion
- These assemblies are no longer copied to the output folder.
- In rare cases, the additional APIs in the default load set might cause name or type conflicts. To resolve a conflict, add more explicit `using` directives, use an alias, or use a fully qualified type name.
```
